### ISSUE TITLE
Aggregation not getting merged

### ISSUE THREAD

**Issue by shweppsie** [jacksontj/promxy#311]
We have 2 identical Prometheus servers. One morning someone broke access from one of the nodes so it was missing data from that morning onwards. An instance query correctly started returning data from the secondary node as it was absent on the primary node. However an aggregation query was only fulfilled on the primary node which meant the results were wrong as they only reflect the data available from that morning.
Query: `max_over_time(solar_charge_state[18h])`
We are still in the process of moving over from thanos to promxy (with victoriametrics) and thanos seemingly manages to return the correct value. Is this a misunderstanding on my part about how promxy works, or is this a bug?

Config:
```
promxy:
  server_groups:
    - static_configs:
        - targets:
          - 10.6.0.5:19090
          - 10.6.2.4:19090
          labels:
            sg: 1
          anti_affinity: 10s
```

**Comment by jacksontj** [comment]
This *sounds* like a but, but its also possible there is just some weird interaction (maybe your scrape interval is super short or something). Would you be able to provide the trace level logs from promxy for this query (query where you expect it to merge but it isn't)?

**Comment by shweppsie** [comment]
Sure thing. I've narrowed down the query to a single controller to reduce the data I have to paste but the problem persists the same.

Here's the query:
```
max_over_time(solar_charge_state{instance="rt1.scot.lw.rur",device="pci-0000:00:15.0-usb-0:3.1:1.0-port0"}[18h])
```

I'm running it at timestamp: 2020-05-22 08:00:00

Here's the results on the prometheus node missing data:
```
{"status":"success","data":{"resultType":"vector","result":[{"metric":{"device":"pci-0000:00:15.0-usb-0:3.1:1.0-port0","instance":"rt1.scot.lw.rur","job":"lw_pwr"},"value":[1590134400,"5"]}]}}
```
and the other with full data
```
{"status":"success","data":{"resultType":"vector","result":[{"metric":{"device":"pci-0000:00:15.0-usb-0:3.1:1.0-port0","instance":"rt1.scot.lw.rur","job":"lw_pwr"},"value":[1590134400,"7"]}]}}
```

Query as sent to promxy.
```
http://10.6.0.5:9090/api/v1/query?query=max_over_time(solar_charge_state%7Binstance=%22rt1.scot.lw.rur%22,device=%22pci-0000:00:15.0-usb-0:3.1:1.0-port0%22%7D%5B18h%5D)&time=2020-05-22T08:00:00Z
```

Here's the debug from promxy.
```
time="2020-05-23T02:34:11Z" level=debug msg="call max_over_time(solar_charge_state{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\",instance=\"rt1.scot.lw.rur\"}[18h]) vector"
time="2020-05-23T02:34:11Z" level=debug msg="http://10.6.2.4:19090" api=Query query="max_over_time(solar_charge_state{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\",instance=\"rt1.scot.lw.rur\"}[18h])" ts="2020-05-22 08:00:00 +0000 UTC"
time="2020-05-23T02:34:11Z" level=debug msg="http://10.6.0.5:19090" api=Query query="max_over_time(solar_charge_state{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\",instance=\"rt1.scot.lw.rur\"}[18h])" ts="2020-05-22 08:00:00 +0000 UTC"
time="2020-05-23T02:34:11Z" level=trace msg="http://10.6.0.5:19090" api=Query error="<nil>" query="max_over_time(solar_charge_state{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\",instance=\"rt1.scot.lw.rur\"}[18h])" took=6.702633ms ts="2020-05-22 08:00:00 +0000 UTC" value="{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\", instance=\"rt1.scot.lw.rur\", job=\"lw_pwr\", sg=\"1\"} => 5 @[1590134400]" warnings="[]"
time="2020-05-23T02:34:11Z" level=trace msg="http://10.6.2.4:19090" api=Query error="<nil>" query="max_over_time(solar_charge_state{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\",instance=\"rt1.scot.lw.rur\"}[18h])" took=8.058252ms ts="2020-05-22 08:00:00 +0000 UTC" value="{device=\"pci-0000:00:15.0-usb-0:3.1:1.0-port0\", instance=\"rt1.scot.lw.rur\", job=\"lw_pwr\", sg=\"1\"} => 7 @[1590134400]" warnings="[]"
192.168.0.135 - - [23/May/2020 02:34:11] "GET /api/v1/query HTTP/1.1 200 181" 0.008831 query=max_over_time%28solar_charge_state%7Binstance%3D%22rt1.scot.lw.rur%22%2Cdevice%3D%22pci-0000%3A00%3A15.0-usb-0%3A3.1%3A1.0-port0%22%7D%5B18h%5D%29&time=2020-05-22T08%3A00%3A00Z
```

Prometheus job config
```
  - file_sd_configs:
    - files:
      - /etc/prometheus/static_configs/lw_pwr.yml
    job_name: lw_pwr
    metrics_path: /metrics
    scrape_interval: 60s
    scrape_timeout: 50s
```

Just to clarify how promxy works, does it get all the data and then run the aggregation locally? So in my case I'd anticipate it would run `solar_charge_state[20h]` on both prometheus instances and then do the `max_over_time` locally to work over any gaps in the source data. Looking at the output it seems like it runs the full max_over_time query on each node and then just picks the first reply.

Here's graphs of the source data in case I'm being unclear.
![image](https://user-images.githubusercontent.com/68794/82720180-11f6b680-9d05-11ea-856b-a32c15467b47.png)
![image](https://user-images.githubusercontent.com/68794/82720082-59c90e00-9d04-11ea-85ce-0a9024b3cd94.png)

**Comment by jacksontj** [comment]
> Just to clarify how promxy works, does it get all the data and then run the aggregation locally? So in my case I'd anticipate it would run solar_charge_state[20h] on both prometheus instances and then do the max_over_time locally to work over any gaps in the source data. Looking at the output it seems like it runs the full max_over_time query on each node and then just picks the first reply.

Not quite; this was actually the initial implementation (fetch all raw data and do math on promxy) but the performance of this is abysmal. The reason it was so slow was that the amount of data we'd need to pass for raw calculations is quite large (in practice the data size just caused huge marshal/unmarshal times). So what promxy actually does is use a `NodeReplacer` feature to effectively break the query into subqueries to farm out to the downstream prom API endpoints. For example, if you want to know the `sum(X)` I don't need to get all the raw data, I can just sum the results of `sum(X)` from both downstream. All the logic for this replacement is found [here](https://github.com/jacksontj/promxy/blob/master/pkg/proxystorage/proxy.go#L179).

So in your case the incoming query (`max_over_time(solar_charge_state{device="pci-0000:00:15.0-usb-0:3.1:1.0-port0",instance="rt1.scot.lw.rur"}[18h])`) is sent to the downstream nodes (as seen in the logs). So the issue we have here is that the raw data has a hole, but this subquery doesn't have one (since its calculating something over-time). 

I have hacked up a PoC of what a fix would need to do (its a terrible hack that I spent ~30m on, so definitely will need better work to replace it :) ) -- https://github.com/jacksontj/promxy/pull/312


So TLDR; not desired behavior, but not entirely unexpected -- fix is somewhat non-trivial (but definitely possible).

**Comment by shweppsie** [comment]
Aha! Thanks for the comprehesive explanation; Much appreciated. I'll test that out tomorrow and just confirm it does fix the issue. If there's anything I can do to help get that code in a state you're happy with, I'm happy to do what I can.

**Comment by anubhav21sharma** [comment]
@jacksontj I see the hack you used to fix this is to divide the range into 10 smaller ranges so that _some_ holes can be detected from the query results and the merged data is closer to the correct answer.

Do you have a better fix in mind for this? The only thing I can think of is to fetch raw results like Thanos does, but that has big performance implications.

**Comment by jacksontj** [comment]
@anubhav21sharma The hack above is actually a bit different. So in general promxy aggregates by sending query or query_range downstream (depending on what query came in). These results are "merged" based on holes in the result data. This dramatically reduces the data that needs to be sent between promxy and the downstream prometheus API (which is why the perf with promxy is significantly better than thanos) but it does mean (1) we are relying on the remote to properly calculate the value and (2) if data is partially missing on one of the downstreams you can get some not-quite accurate data (e.g. if you have sum(foo) but 1 of the downstreams was missing 1 timeseries its result would not have holes but not be correct, this failure mode is pretty uncommon, but does exist.)

The hack above is to support over_time aggregation methods, which basically needs re-evaluation over a different range with a different step etc. The "hack" above doesn't do it properly (meaning it only gets the correct result for a few situations) in reality what I need to do is pass the Engine object down to this level so it can recurse on itself and get the correct result.

**Comment by jacksontj** [comment]
This issue has been resolved! Apparently there is a bit of magic in how offset ranges are calculated with subqueries; I fixed that in #381 and with that this issue is resolved!
